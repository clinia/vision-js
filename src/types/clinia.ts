import { SearchClient as SearchClientV2 } from 'clinia';
import { SearchResponse as SearchResponseV2 } from '@clinia/client-search';

export type Client = SearchClientV2;

export interface MultiResponse<TRecord = any> {
  results: Array<SearchResponse<TRecord>>;
}

export type SearchResponse<TRecord> = SearchResponseV2<TRecord>;
