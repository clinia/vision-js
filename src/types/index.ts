export * from './clinia';
export * from './connector';
export * from './insights';
export * from './vision';
export * from './widget';
