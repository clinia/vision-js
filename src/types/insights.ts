export type InsightsClientMethod =
  | 'clickedRecordIDsAfterSearch'
  | 'convertedRecordIDsAfterSearch';

export type InsightsClientPayload = {
  eventName: string;
  queryID: string;
  index: string;
  ids: string[];
  positions?: number[];
};

export type InsightsClient = (
  method: InsightsClientMethod,
  payload: InsightsClientPayload
) => void;

export type InsightsClientWrapper = (
  method: InsightsClientMethod,
  payload: Partial<InsightsClientPayload>
) => void;
