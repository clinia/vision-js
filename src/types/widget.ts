import { Index } from '../widgets/index/index';
import {
  CliniaSearchHelper as Helper,
  SearchParameters,
  SearchResults,
  PlainSearchParameters,
} from '@clinia/search-helper';
import { Vision } from './vision';

export interface InitOptions {
  visionInstance: Vision;
  parent: Index | null;
  uiState: UiState;
  state: SearchParameters;
  helper: Helper;
  templatesConfig: object;
  createURL(state: SearchParameters): string;
}

export interface ScopedResult {
  indexId: string;
  results: SearchResults;
  helper: Helper;
}

export interface RenderOptions {
  visionInstance: Vision;
  templatesConfig: object;
  results: SearchResults;
  scopedResults: ScopedResult[];
  state: SearchParameters;
  helper: Helper;
  searchMetadata: {
    isSearchStalled: boolean;
  };
  createURL(state: SearchParameters): string;
}

export interface DisposeOptions {
  helper: Helper;
  state: SearchParameters;
}

export interface WidgetStateOptions {
  searchParameters: SearchParameters;
  helper: Helper;
}

export interface WidgetSearchParametersOptions {
  uiState: IndexUiState;
}

export type IndexUiState = {
  query?: string;
  refinementList?: {
    [property: string]: string[];
  };
  menu?: {
    [property: string]: string;
  };

  toggle?: {
    [property: string]: boolean;
  };
  geoSearch?: {
    /**
     * The rectangular area in geo coordinates.
     * The rectangle is defined by two diagonally opposite points, hence by 4 floats separated by commas.
     *
     * @example '47.3165,4.9665,47.3424,5.0201'
     */
    boundingBox: string;
  };
  page?: number;
  perPage?: number;
  configure?: PlainSearchParameters;
  places?: {
    query: string;
    /**
     * The central geolocation.
     *
     * @example '48.8546,2.3477'
     */
    position: string;
  };
};

export type UiState = {
  [indexId: string]: IndexUiState;
};

/**
 * Widgets are the building blocks of Visionjs. Any valid widget must
 * have at least a `render` or a `init` function.
 */
export interface Widget {
  $$type?:
    | 'cvi.autocomplete'
    | 'cvi.breadcrumb'
    | 'cvi.clearRefinements'
    | 'cvi.configure'
    | 'cvi.configureRelatedItems'
    | 'cvi.currentRefinements'
    | 'cvi.geoSearch'
    | 'cvi.hierarchicalMenu'
    | 'cvi.records'
    | 'cvi.perPage'
    | 'cvi.index'
    | 'cvi.infiniteRecords'
    | 'cvi.menu'
    | 'cvi.numericMenu'
    | 'cvi.pagination'
    | 'cvi.places'
    | 'cvi.poweredBy'
    | 'cvi.queryRules'
    | 'cvi.queryRuleCustomData'
    | 'cvi.queryRuleContext'
    | 'cvi.range'
    | 'cvi.rangeInput'
    | 'cvi.rangeSlider'
    | 'cvi.ratingMenu'
    | 'cvi.refinementList'
    | 'cvi.searchBox'
    | 'cvi.sortBy'
    | 'cvi.stats'
    | 'cvi.toggleRefinement'
    | 'cvi.voiceSearch';
  /**
   * Called once before the first search
   */
  init?(options: InitOptions): void;
  /**
   * Called after each search response has been received
   */
  render?(options: RenderOptions): void;
  /**
   * Called when this widget is unmounted. Used to remove refinements set by
   * during this widget's initialization and life time.
   */
  dispose?(options: DisposeOptions): SearchParameters | void;
  /**
   * This function is required for a widget to be taken in account for routing.
   * It will derive a uiState for this widget based on the existing uiState and
   * the search parameters applied.
   * @param uiState current state
   * @param widgetStateOptions extra information to calculate uiState
   */
  getWidgetState?(
    uiState: IndexUiState,
    widgetStateOptions: WidgetStateOptions
  ): IndexUiState;
  /**
   * This function is required for a widget to behave correctly when a URL is
   * loaded via e.g. routing. It receives the current UiState and applied search
   * parameters, and is expected to return a new search parameters.
   * @param state applied search parameters
   * @param widgetSearchParametersOptions extra information to calculate next searchParameters
   */
  getWidgetSearchParameters?(
    state: SearchParameters,
    widgetSearchParametersOptions: WidgetSearchParametersOptions
  ): SearchParameters;
}

export type WidgetFactory<TWidgetParams> = (
  widgetParams: TWidgetParams
) => Widget;

export type Template<TTemplateData = void> =
  | string
  | ((data: TTemplateData) => string);
