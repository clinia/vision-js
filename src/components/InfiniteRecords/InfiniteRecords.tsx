/** @jsx h */

import { h } from 'preact';
import cx from 'classnames';
import Template from '../Template/Template';
import { SearchResults } from '@clinia/search-helper';
import { Records } from '../../types';
import { InfiniteRecordsTemplates } from '../../widgets/infinite-records/infinite-records';

type InfiniteRecordsCSSClasses = {
  root: string;
  emptyRoot: string;
  list: string;
  item: string;
  loadPrevious: string;
  disabledLoadPrevious: string;
  loadMore: string;
  disabledLoadMore: string;
};

type InfiniteRecordsProps = {
  cssClasses: InfiniteRecordsCSSClasses;
  records: Records;
  results: SearchResults;
  hasShowPrevious: boolean;
  showPrevious: () => void;
  showMore: () => void;
  templateProps: {
    [key: string]: any;
    templates: InfiniteRecordsTemplates;
  };
  isFirstPage: boolean;
  isLastPage: boolean;
};

const InfiniteRecords = ({
  results,
  records,
  hasShowPrevious,
  showPrevious,
  showMore,
  isFirstPage,
  isLastPage,
  cssClasses,
  templateProps,
}: InfiniteRecordsProps) => {
  if (results.records.length === 0) {
    return (
      <Template
        {...templateProps}
        templateKey="empty"
        rootProps={{
          className: cx(cssClasses.root, cssClasses.emptyRoot),
        }}
        data={results}
      />
    );
  }

  return (
    <div className={cssClasses.root}>
      {hasShowPrevious && (
        <Template
          {...templateProps}
          templateKey="showPreviousText"
          rootTagName="button"
          rootProps={{
            className: cx(cssClasses.loadPrevious, {
              [cssClasses.disabledLoadPrevious]: isFirstPage,
            }),
            disabled: isFirstPage,
            onClick: showPrevious,
          }}
        />
      )}

      <ol className={cssClasses.list}>
        {records.map((record, position) => (
          <Template
            {...templateProps}
            templateKey="item"
            rootTagName="li"
            rootProps={{ className: cssClasses.item }}
            key={record.id}
            data={{
              ...record,
              __recordIndex: position,
            }}
          />
        ))}
      </ol>

      <Template
        {...templateProps}
        templateKey="showMoreText"
        rootTagName="button"
        rootProps={{
          className: cx(cssClasses.loadMore, {
            [cssClasses.disabledLoadMore]: isLastPage,
          }),
          disabled: isLastPage,
          onClick: showMore,
        }}
      />
    </div>
  );
};

export default InfiniteRecords;
