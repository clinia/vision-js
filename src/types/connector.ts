import { SearchResults } from '@clinia/search-helper';
import { Records, Vision } from './vision';
import { InsightsClient } from './insights';

export type RendererOptions<TWidgetParams = unknown> = {
  /**
   * Original parameters for this widget.
   * Useful for giving back the render parameters to the renderer.
   */
  widgetParams: TWidgetParams;
  visionInstance: Vision;
  results?: SearchResults;
  records?: Records;
  insights?: InsightsClient;
};

export type Renderer<TRenderOptions extends RendererOptions = any> = (
  renderOptions: TRenderOptions,
  isFirstRender: boolean
) => void;

export type Unmounter = () => void;
