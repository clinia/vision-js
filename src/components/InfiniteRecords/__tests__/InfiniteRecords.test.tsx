/** @jsx h */

import { h } from 'preact';
import { render } from '@testing-library/preact';
import { SearchResults } from '@clinia/search-helper';
import InfiniteRecords from '../InfiniteRecords';
import { Records } from '../../../types';

describe('InfiniteRecrods', () => {
  const cssClasses = {
    root: 'root',
    emptyRoot: 'emptyRoot',
    item: 'item',
    list: 'list',
    loadPrevious: 'loadPrevious',
    disabledLoadPrevious: 'disabledLoadPrevious',
    loadMore: 'loadMore',
    disabledLoadMore: 'disabledLoadMore',
  };

  describe('markup', () => {
    it('should render <InfiniteRecords /> on first page', () => {
      const records: Records = [
        {
          id: 'one',
          foo: 'bar',
          __position: 0,
        },
        {
          id: 'two',
          foo: 'baz',
          __position: 1,
        },
      ];

      const props = {
        hasShowPrevious: false,
        showPrevious: () => {},
        showMore: () => {},
        results: { records },
        records,
        isFirstPage: true,
        isLastPage: false,
        templateProps: {
          templates: {
            empty: 'empty',
            showPreviousText: 'showPreviousText',
            showMoreText: 'showMoreText',
            item: 'item',
          },
        },
        cssClasses,
      };

      const { container } = render(<InfiniteRecords {...props} />);

      expect(container).toMatchSnapshot();
    });

    it('should render <InfiniteRecords /> on last page', () => {
      const records: Records = [
        {
          id: 'one',
          foo: 'bar',
          __position: 0,
        },
        {
          id: 'two',
          foo: 'baz',
          __position: 1,
        },
      ];

      const props = {
        hasShowPrevious: false,
        showPrevious: () => {},
        showMore: () => {},
        results: { records } as SearchResults,
        records,
        isFirstPage: false,
        isLastPage: true,
        templateProps: {
          templates: {
            empty: 'empty',
            showPreviousText: 'showPreviousText',
            showMoreText: 'showMoreText',
            item: 'item',
          },
        },
        cssClasses,
      };

      const { container } = render(<InfiniteRecords {...props} />);

      expect(container).toMatchSnapshot();
    });

    it('should render <InfiniteRecords /> without records on first page', () => {
      const records: Records = [];

      const props = {
        hasShowPrevious: false,
        showPrevious: () => {},
        showMore: () => {},
        results: { records } as SearchResults,
        records,
        isFirstPage: true,
        isLastPage: false,
        templateProps: {
          templates: {
            empty: 'empty',
            showPreviousText: 'showPreviousText',
            showMoreText: 'showMoreText',
            item: 'item',
          },
        },
        cssClasses,
      };

      const { container } = render(<InfiniteRecords {...props} />);

      expect(container).toMatchSnapshot();
    });

    it('should render <InfiniteRecords /> without records on last page', () => {
      const records: Records = [];

      const props = {
        hasShowPrevious: false,
        showPrevious: () => {},
        showMore: () => {},
        results: { records } as SearchResults,
        records,
        isFirstPage: false,
        isLastPage: true,
        templateProps: {
          templates: {
            empty: 'empty',
            showPreviousText: 'showPreviousText',
            showMoreText: 'showMoreText',
            item: 'item',
          },
        },
        cssClasses,
      };

      const { container } = render(<InfiniteRecords {...props} />);

      expect(container).toMatchSnapshot();
    });

    it('should render <InfiniteRecords /> with "Show previous" button on first page', () => {
      const records: Records = [
        {
          id: 'one',
          foo: 'bar',
          __position: 0,
        },
        {
          id: 'two',
          foo: 'baz',
          __position: 1,
        },
      ];

      const props = {
        hasShowPrevious: true,
        showPrevious: () => {},
        showMore: () => {},
        results: { records } as SearchResults,
        records,
        isFirstPage: true,
        isLastPage: false,
        templateProps: {
          templates: {
            empty: 'empty',
            showPreviousText: 'showPreviousText',
            showMoreText: 'showMoreText',
            item: 'item',
          },
        },
        cssClasses,
      };

      const { container } = render(<InfiniteRecords {...props} />);

      const previousButton = container.querySelector('.loadPrevious');

      expect(previousButton).toBeInTheDocument();
      expect(previousButton).toHaveClass('disabledLoadPrevious');
      expect(previousButton).toBeDisabled();
      expect(container).toMatchSnapshot();
    });

    it('should render <InfiniteRecords /> with "Show previous" button on last page', () => {
      const records: Records = [
        {
          id: 'one',
          foo: 'bar',
          __position: 0,
        },
        {
          id: 'two',
          foo: 'baz',
          __position: 1,
        },
      ];

      const props = {
        hasShowPrevious: true,
        showPrevious: () => {},
        showMore: () => {},
        results: { records } as SearchResults,
        records,
        isFirstPage: false,
        isLastPage: true,
        templateProps: {
          templates: {
            empty: 'empty',
            showPreviousText: 'showPreviousText',
            showMoreText: 'showMoreText',
            item: 'item',
          },
        },
        cssClasses,
      };

      const { container } = render(<InfiniteRecords {...props} />);

      const previousButton = container.querySelector('.loadPrevious');

      expect(previousButton).toBeInTheDocument();
      expect(previousButton).not.toHaveClass('disabledLoadPrevious');
      expect(previousButton).not.toBeDisabled();
      expect(container).toMatchSnapshot();
    });
  });
});
