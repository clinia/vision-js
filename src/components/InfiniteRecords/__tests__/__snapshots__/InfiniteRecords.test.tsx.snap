// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteRecrods markup should render <InfiniteRecords /> on first page 1`] = `
<div>
  <div
    class="root"
  >
    <ol
      class="list"
    >
      <li
        class="item"
      >
        item
      </li>
      <li
        class="item"
      >
        item
      </li>
    </ol>
    <button
      class="loadMore"
    >
      showMoreText
    </button>
  </div>
</div>
`;

exports[`InfiniteRecrods markup should render <InfiniteRecords /> on last page 1`] = `
<div>
  <div
    class="root"
  >
    <ol
      class="list"
    >
      <li
        class="item"
      >
        item
      </li>
      <li
        class="item"
      >
        item
      </li>
    </ol>
    <button
      class="loadMore disabledLoadMore"
      disabled=""
    >
      showMoreText
    </button>
  </div>
</div>
`;

exports[`InfiniteRecrods markup should render <InfiniteRecords /> with "Show previous" button on first page 1`] = `
<div>
  <div
    class="root"
  >
    <button
      class="loadPrevious disabledLoadPrevious"
      disabled=""
    >
      showPreviousText
    </button>
    <ol
      class="list"
    >
      <li
        class="item"
      >
        item
      </li>
      <li
        class="item"
      >
        item
      </li>
    </ol>
    <button
      class="loadMore"
    >
      showMoreText
    </button>
  </div>
</div>
`;

exports[`InfiniteRecrods markup should render <InfiniteRecords /> with "Show previous" button on last page 1`] = `
<div>
  <div
    class="root"
  >
    <button
      class="loadPrevious"
    >
      showPreviousText
    </button>
    <ol
      class="list"
    >
      <li
        class="item"
      >
        item
      </li>
      <li
        class="item"
      >
        item
      </li>
    </ol>
    <button
      class="loadMore disabledLoadMore"
      disabled=""
    >
      showMoreText
    </button>
  </div>
</div>
`;

exports[`InfiniteRecrods markup should render <InfiniteRecords /> without records on first page 1`] = `
<div>
  <div
    class="root emptyRoot"
  >
    empty
  </div>
</div>
`;

exports[`InfiniteRecrods markup should render <InfiniteRecords /> without records on last page 1`] = `
<div>
  <div
    class="root emptyRoot"
  >
    empty
  </div>
</div>
`;
